src/debug/showUserInput.ts

```diff
@@ -0,0 +1,212 @@
+          fill: black;
+          transition: none;
+        }
+      </style>
+      <svg width="40" height="40">
+        <polygon points="0,0 0,34 8,28 14,38 20,35 14,26 24,24"/>
+      </svg>`;
+      element.style.position = 'fixed';
+      element.style.pointerEvents = 'none';
+      element.style.zIndex = String(Number.MAX_SAFE_INTEGER - 1);
+      document.documentElement.appendChild(element);
+      return element;
+    }
+  }, args);
+}
+
+const textForPage = new WeakMap<Page, string>();
+async function updateKeyboardIndicator(page: Page, key: string, code: string, modifiers: Set<types.KeyboardModifier>) {
+  let text;
+  let dim = true;
+  if (!key) {
+    if (!textForPage.get(page))
+      return;
+    textForPage.delete(page);
+    text = '';
+  } else if (!modifiers.size && key.trim().length === 1) {
+    const currentText = (textForPage.get(page) || '') + key;
+    textForPage.set(page, currentText);
+    text = currentText;
+    dim = false;
+  } else {
+    text = code;
+    textForPage.delete(page);
+  }
+  if (text) {
+    // If ControlLeft is pressed, Control+ControlLeft would be redundant.
+    const filteredModifiers = [...modifiers].filter(modifier => !code.includes(modifier));
+    text = [...filteredModifiers, text].join(' + ');
+  }
+  await renderKeyboardIndicator(page, text, dim);
+}
+
+async function renderKeyboardIndicator(page: Page, text: string, dim = false) {
+  await evaluateSafelyInTheUtilityContext(page, ({text, dim}) => {
+    const element: HTMLElement = document.querySelector(':scope > playwright-keyboard-indicator') || createKeyboardIndicator();
+    (element as any).__setText(text, dim);
+
+    function createKeyboardIndicator() {
+      const element = document.createElement('playwright-keyboard-indicator');
+      const shadow = element.attachShadow({mode: 'closed'});
+      (element as any).__shadowForTest = shadow;
+      shadow.innerHTML = `
+      <style>
+        :host {
+          position: fixed;
+          bottom: 0;
+          left: 0;
+          right: 0;
+          z-index: ${Number.MAX_SAFE_INTEGER - 1};
+          pointer-events: none;
+          margin: 10px;
+          text-align: center;
+          user-select: none;
+        }
+        div {
+          color: white;
+          font-family: sans-serif;
+          font-size: 60px;
+          background: rgba(0, 0, 0, 0.75);
+          display: inline-block;
+          border-radius: 10px;
+          padding: 10px;
+        }
+        div.dim {
+          color: #73C2FB;
+        }
+      </style>`;
+      const div = document.createElement('div');
+      let timeout: any = null;
+      (element as any).__setText = (text: string, dim: boolean) => {
+        if (text) {
+          div.textContent = text;
+          shadow.appendChild(div);
+        }
+        if (timeout)
+          clearInterval(timeout);
+
+        timeout = setTimeout(() => {
+          div.remove();
+          timeout = null;
+        }, 200);
```

Comment:
Played with this, 200ms is too short for me. WDYT about reusing the slowMo value is present? Same for tap.

Reply:
Sure. `(slowMo || 150) + 50`

Reply:
For tap I think always 500ms is fine. We treat tap as an atomic action and dont slowMo inbetween touchstart and touchend.